pandas/core/algorithms.py

```diff
@@ -65,6 +65,8 @@ def _ensure_data(values, dtype=None):
 
     # we check some simple dtypes first
     try:
+        if is_object_dtype(dtype):
```

Comment:
Not quite: the existing check is for the `values`.  This is for the `dtype` specified.  This is essentially an `O(1)` operation.

And yes, I did check performance, and I didn't see any issues on my machine.